compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1097,76 +1096,159 @@ private BType inferTableMemberType(List<BType> memTypes, BType expType) {
-                field.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED)) + Flags.asMask(EnumSet.of(Flag.READONLY));
-            } else if (requiredFieldNames.contains(fieldName)) {
-                field.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED));
+                resultantField.symbol.flags = Flags.OPTIONAL;
+            } else if (keySpecifierFieldNames.contains(fieldName)) {
+                resultantField.symbol.flags = Flags.REQUIRED | Flags.READONLY;
+            } else {
+                resultantField.symbol.flags = Flags.REQUIRED;
             }
+
+            inferredFields.add(resultantField);
         }
 
-        return createTableConstraintRecordType(allFieldSet, restFieldTypes, tableConstructorExpr.pos);
+        return createTableConstraintRecordType(inferredFields, restFieldTypes, tableConstructorExpr.pos);
     }
 
-    private BRecordType createTableConstraintRecordType(Set<BField> allFieldSet, List<BType> restFieldTypes,
+    /**
+     * Create a new {@code BField} out of existing {@code BField}, while changing its type.
+     * The new type is derived from the given list of bTypes.
+     *
+     * @param field  - existing {@code BField}
+     * @param bTypes - list of bTypes
+     * @return a {@code BField}
+     */
+    private BField createFieldWithType(BField field, List<BType> bTypes) {
+        BType resultantType = getResultantUnionOrNonUnionType(bTypes);
+
+        BVarSymbol originalSymbol = field.symbol;
+        BVarSymbol fieldSymbol = new BVarSymbol(originalSymbol.flags, originalSymbol.name, originalSymbol.pkgID,
+                resultantType, originalSymbol.owner, originalSymbol.pos, VIRTUAL);
+
+        return new BField(field.name, field.pos, fieldSymbol);
+    }
+
+    /**
+     * Get the resultant union or non-union type from a {@code List<BType>}.
+     *
+     * @param bTypes bType list (size > 0)
+     * @return {@code BUnionType} if effective members in list is > 1. {@code BType} Otherwise.
+     */
+    private BType getResultantUnionOrNonUnionType(List<BType> bTypes) {
+        if (bTypes.size() == 1) {
+            return bTypes.get(0);
+        }
+
+        LinkedHashSet<BType> bTypeSet = new LinkedHashSet<>(bTypes);
+        List<BType> eBTypes = new ArrayList<>(bTypes.size());
+        addEffectiveMemberTypes(eBTypes, bTypeSet);
+
+        return getRepresentativeBroadType(eBTypes);
+    }
+
+    private void addEffectiveMemberTypes(List<BType> eBTypes, LinkedHashSet<BType> bTypes) {
+        for (BType memberType : bTypes) {
+            BType bType;
+            switch (memberType.tag) {
+                case TypeTags.NEVER:
+                    continue;
+                case TypeTags.UNION:
+                    addEffectiveMemberTypes(eBTypes, ((BUnionType) memberType).getMemberTypes());
+                    continue;
+                case TypeTags.TYPEREFDESC:
+                    BType constraint = types.getReferredType(memberType);
```

Comment:
Doesn't `getRepresentativeBroadType` handle these as well? For type reference types we should probably update `getRepresentativeBroadType` itself to use `types.getReferredType(inferredTypeList.get(i))` as the `type` instead of `inferredTypeList.get(i)`.

Reply:
But getRepresentativeBroadType does not do the flattening right?
e.g.
```bal
function testTableConstructorWithVar5() {
    int|never f = 1;

    var v1 = table [
            {a: f},
            {a: 0}
        ];
    int _ = v1;
}
```
`ERROR [test_parser.bal:(8:13,8:15)] incompatible types: expected 'int', found 'table<record {| (int|never) a; |}>'`
Basically what it does is eliminate a type if it is assignable to another type.

Also for type references `inferredTypeList.get(i)` and `types.getReferredType(inferredTypeList.get(i))` have no difference to the `types.isAssignable(otherType, type)` right? We have to loop again and modify the list replacing types with `types.getReferredType(inferredTypeList.get(i))`?

Reply:
I think the issue with `int|never` and `int` not being simplified to `int` in `getRepresentativeBroadType` is because we are passing unions as is right? 

What if for each member type `T` in the list passed as a parameter to `getResultantUnionOrNonUnionType`, we create a new set of members where

1. if `T` is a type reference type we get the referred type and use that as `T`
2. if `T` is a union add to the set of members the result of calling `org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType#getMemberTypes`, if not we add `T` as is

The reason to call `getReferredType` is for this flattening, so it has to happen here (point 1.) instead of in `getRepresentativeBroadType` like I originally suggested.